provider/openstack/provider_test.go

```diff
@@ -361,6 +363,141 @@ func (s *localTests) detectRegions(c *gc.C) ([]cloud.Region, error) {
 	return provider.(environs.CloudRegionDetector).DetectRegions()
 }
 
+func (s *localTests) TestSchema(c *gc.C) {
+	y := []byte(`
+auth-types: [userpass, access-key]
+endpoint: http://foo.com/openstack
+regions: 
+  one:
+    endpoint: http://foo.com/bar
+  two:
+    endpoint: http://foo2.com/bar2
+`[1:])
+	var v interface{}
+	err := yaml.Unmarshal(y, &v)
+	c.Assert(err, jc.ErrorIsNil)
+	v, err = utils.ConformYAML(v)
+	c.Assert(err, jc.ErrorIsNil)
+
+	p, err := environs.Provider("openstack")
+	err = p.CloudSchema().Validate(v)
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (localTests) TestPingInvalidHost(c *gc.C) {
+	tests := []string{
+		"foo.com",
+		"http://IHopeNoOneEverBuysThisVerySpecificJujuDomainName.com",
+		"http://IHopeNoOneEverBuysThisVerySpecificJujuDomainName:77",
+	}
+
+	p, err := environs.Provider("openstack")
+	c.Assert(err, jc.ErrorIsNil)
+	for _, t := range tests {
+		err = p.Ping(t)
+		if err == nil {
+			c.Errorf("ping %q: expected error, but got nil.", t)
+			continue
+		}
+		expected := "No Openstack server running at " + t
+		if err.Error() != expected {
+			c.Errorf("ping %q: expected %q got %v", t, expected, err)
+		}
+	}
+}
+func (localTests) TestPingNoEndpoint(c *gc.C) {
+	server := httptest.NewServer(http.HandlerFunc(http.NotFound))
+	defer server.Close()
+	p, err := environs.Provider("openstack")
+	c.Assert(err, jc.ErrorIsNil)
+	err = p.Ping(server.URL)
+	c.Assert(err, gc.ErrorMatches, "No Openstack server running at "+server.URL)
+}
+
+func (localTests) TestPingInvalidResponse(c *gc.C) {
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprint(w, "Hi!")
+	}))
+	defer server.Close()
+	p, err := environs.Provider("openstack")
+	c.Assert(err, jc.ErrorIsNil)
+	err = p.Ping(server.URL)
+	c.Assert(err, gc.ErrorMatches, "No Openstack server running at "+server.URL)
+}
+
+func (localTests) TestPingOK(c *gc.C) {
```

Comment:
Is there any way we can turn this into a unit test? We can't climb out of a hole by digging.